src/test/java/com/endava/internship/collections/GenericMapTest.java

```diff
@@ -0,0 +1,340 @@
+     * middle element of the bucket, the last element in the bucket).
+     * Map should ensure removing those elements in any test case
+     * * ------------------------- *
+     */
+    @ParameterizedTest
+    @MethodSource("provideObjectListWithTheSameHashIndex")
+    public void shouldRemoveFirstObjectInTheBucket(List<String> objectList){
+        for (String object : objectList) {
+            genericMap.put(object, 1);
+        }
+        //removing first element in the bucket
+        String firstObjectInTheBucket = "string 1";
+        genericMap.remove(firstObjectInTheBucket);
+
+        assertAll(
+                () -> assertThat(genericMap.size()).isEqualTo(objectList.size() - 1),
+                () -> assertThat(genericMap.containsKey(firstObjectInTheBucket)).isFalse()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideObjectListWithTheSameHashIndex")
+    public void shouldRemoveMiddleObjectInTheBucket(List<String> objectList){
+        for (String object : objectList) {
+            genericMap.put(object, 1);
+        }
+        //removing middle element in the bucket
+        String middleObjectInTheBucket = "string 10";
+        genericMap.remove(middleObjectInTheBucket);
+
+        assertAll(
+                () -> assertThat(genericMap.size()).isEqualTo(objectList.size() - 1),
+                () -> assertThat(genericMap.containsKey(middleObjectInTheBucket)).isFalse()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideObjectListWithTheSameHashIndex")
+    public void shouldRemoveLastObjectInTheBucket(List<String> objectList){
+        for (String object : objectList) {
+            genericMap.put(object, 1);
+        }
+        //removing last element in the bucket
+        String lastObjectInTheBucket = "string 21";
+        genericMap.remove(lastObjectInTheBucket);
+
+        assertAll(
+                () -> assertThat(genericMap.size()).isEqualTo(objectList.size() - 1),
+                () -> assertThat(genericMap.containsKey(lastObjectInTheBucket)).isFalse()
+        );
+    }
+
+    public static Stream<Arguments> provideObjectListWithTheSameHashIndex(){
+        List<String> objectList = new ArrayList<>();
+
+        //if map has default initial capacity and has not been rehashed,
+        //all three objects should produce the same hash index (!but not the same hashcode),
+        // and consequently, will be placed in the same bucket in this order.
+        objectList.add("string 1");
+        objectList.add("string 10");
+        objectList.add("string 21");
+
+        return Stream.of(
+                Arguments.of(objectList)
+        );
+    }
+
+    public static Stream<Arguments> provideObjectList(){
+        List<String> objectList = new ArrayList<>();
+        for(int i = 0; i < 5; ++i){
+            objectList.add("string " + i);
+        }
+        return Stream.of(
+                Arguments.of(objectList)
+        );
+
+    }
+}
```

Comment:
Small note, each file should end with an empty line to eliminate this red symbol.